### DISCUSSION TITLE
Is it possible to implement full-fledged inheritance ?

### DISCUSSION THREAD

**Discussion by weedb** [mobxjs/mobx-state-tree#1740] [Q&A] [answered]
Sorry for my english.
I ran into a situation where I needed to override a property in a model during inheritance and types.compose can't do it. I searched all day, but I could not find a reason why inheritance is implemented only in the form types.compose or .props() but it also doesn't work for me. 

So my questions are:
1) Is there any solution that can help me now? (with property overrides)
2) Do you plan to implement regular inheritance?

**Comment by jamonholmgren** [comment, marked as answer]
Your English is excellent @weedb, no problem.

 1. You can redefine the prop. The only caveat is that TypeScript gets a little confused.

https://codesandbox.io/s/thirsty-surf-n7tde?file=/src/App.tsx

As you can see in the above CodeSandbox, I redefined the prop and assigned it a new type. TypeScript doesn't like it, but it works fine at runtime.

 2. No plans to implement regular inheritance. You can look at https://mobx-keystone.js.org/ as an alternative if this is super important to you.

Thanks!

**Reply by weedb** [reply]
Thanks for the answer.

1) In fact, I already tried to compose an example similar to yours, and when I saw the never type, i realized that losing typesctipt power does not suit me.
2) I have already looked at mobx-keystone and for me personally it seems not as good a solution as mst.

But in general, is it difficult to implement regular inheritance for mst? I ask, with the hope that when I have more experience I will try to do it.

**Reply by k-g-a** [reply]
I've already discussed same topic here: https://github.com/mobxjs/mobx-state-tree/issues/1403
In short, I think it should be a runtime error/warning (same as touching dead nodes) to re-define already existing prop via `.porps()` call, but should be definetly possible (with type annotation changes) via special `.overrideProps()` method.

**Reply by jamonholmgren** [reply]
Good idea @k-g-a .
